### PR TITLE
[SQL] Implement SESSION windows

### LINKED ISSUE #6778: [SQL] Implement session windows
This is part of #659 

Session window queries can be rewritten as standard SQL queries using LAG and comparing consecutive timestamps:

```sql
SELECT * FROM TABLE(
  SESSION(TABLE orders, DESCRIPTOR(rowtime), DESCRIPTOR(product), INTERVAL '20' MINUTE));
```

Equivalent form:

```sql
WITH
-- LAG finds each row's predecessor within its key
lagged AS (
  SELECT *,
         LAG(rowtime) OVER (PARTITION BY product ORDER BY rowtime) AS prev
  FROM orders          
  WHERE rowtime IS NOT NULL
),
-- brk = 1 when the row starts a new session
breaks AS (
  SELECT rowtime, id, product, units,
         CASE WHEN prev IS NULL
                OR rowtime >= prev + INTERVAL '20' MINUTE
              THEN 1 ELSE 0 END AS brk
  FROM lagged
),
-- the running sum of breaks numbers each row's session within its key
sessionized AS (
  SELECT rowtime, id, product, units,
         SUM(brk) OVER (PARTITION BY product ORDER BY rowtime
                        RANGE UNBOUNDED PRECEDING) AS sid
  FROM breaks
),
-- one row per session holding its bounds
bounds AS (
  SELECT product, sid, MIN(rowtime) AS min_ts, MAX(rowtime) AS max_ts
  FROM sessionized
  GROUP BY product, sid
)
-- attach the bounds back to every row of the session
SELECT s.rowtime, s.id, s.product, s.units,
       b.min_ts                        AS window_start,
       b.max_ts + INTERVAL '20' MINUTE AS window_end
FROM sessionized s
JOIN bounds b
  ON  s.product IS NOT DISTINCT FROM b.product
  AND s.sid     IS NOT DISTINCT FROM b.sid;
```

### PR DESCRIPTION
**Pull request by mihaibudiu**
Fixes #6778 

The issue describes how this is done. SESSION is a "table function" in SQL, there is a Calcite-level transformation which rewrites it into some window aggregates using LAG and some self-joins. After this rewrite there is nothing really special about session.

The PR is a bit more involved because it uncovered a bug in the handling GC operators in the graph, and it also contains the fix.

(The PR also uncovered 3 bugs in Calcite, which I have sent PRs for, but we don't depend on them.)

### Describe Manual Test Plan

Ran lots of tests.

## Checklist

- [x] Unit tests added/updated
- [x] Documentation updated

### CODE REVIEW

**Review by mythical-fred — APPROVED** [review]
LGTM. The Calcite-level rewrite is clean and well-documented — the Javadoc on `SessionRewriteRule` makes the LAG → break → SUM(break) → self-join plan immediately clear. The `RemoveIdentityOperators` refactor properly handles the GC-conflict edge case that SESSION surfaces, and `StrayGC` now catches duplicate retainers. Test coverage is thorough: keyed and keyless sessions, NULL timestamps and keys, surrounded columns, negative cases, streaming with LATENESS/compaction, deletions, and late merges across the waterline.

**Review by gz — APPROVED** [review]

File: sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/quidem/SessionTests.java [hunk 1]
```diff
@@ -0,0 +1,227 @@
+package org.dbsp.sqlCompiler.compiler.sql.quidem;
```
Comment (gz) [inline sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/quidem/SessionTests.java:1]: whats quidem

File: sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/streaming/StreamingTests.java [hunk 2]
```diff
@@ -869,6 +869,224 @@ public void endVisit() {
...
+                CREATE TABLE events(
+                    ts  TIMESTAMP NOT NULL LATENESS INTERVAL 1 HOURS,
+                    uid VARCHAR
+                );
+                CREATE VIEW sessions AS
+                SELECT uid, COUNT(*) AS cnt, window_start, window_end
+                FROM TABLE(SESSION(TABLE events, DESCRIPTOR(ts), DESCRIPTOR(uid), INTERVAL 10 MINUTES))
+                GROUP BY uid, window_start, window_end;""";
+        CompilerCircuitStream ccs = this.getCCS(sql).compactAfterEachStep();
+        // TODO: window_start cannot have a waterline (a session can be unbounded)
```
Comment (gz) [inline sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/streaming/StreamingTests.java:888]: so does it throw an error in this case?

File: docs.feldera.com/docs/sql/table.md [hunk 3]
```diff
@@ -118,3 +118,55 @@ function:
 - `window_end`, of the same type as the column `orders.rowtime`
 
 A `NULL` timestamp produces no rows in the result.
+
+### `SESSION`
+
+`SESSION` groups rows into sessions based on a timestamp column.  Two
+rows belong to the same session when their timestamps are less than
+`size` (the inactivity gap) apart.  Unlike `TUMBLE` and `HOP` windows,
```
Comment (gz) [inline docs.feldera.com/docs/sql/table.md:126]: > `size` (the inactivity gap) apart.  Unlike `TUMBLE` and `HOP` windows,

what is the inactivity gap

File: docs.feldera.com/docs/sql/table.md [hunk 4]
```diff
@@ -118,3 +118,55 @@ function:
...
+session windows are not fixed in absolute time: each session starts at
+the timestamp of its first row and ends `size` after the timestamp of
+its last row.  The optional `key` descriptor partitions the rows;
+sessions are formed separately within each key.
```
Comment (gz) [inline docs.feldera.com/docs/sql/table.md:130]: it would be nice to have a small example with data so I can quickly parse what this does

**Review by gz — APPROVED** [review]

File: sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/quidem/SessionTests.java (hunk 1 above)
Comment (mihaibudiu) [reply sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/quidem/SessionTests.java:1]: This is how Calcite names some tests, I am reusing them whenever I can. It's Latin for "in fact".

File: docs.feldera.com/docs/sql/table.md (hunk 3 above)
Comment (mihaibudiu) [reply docs.feldera.com/docs/sql/table.md:126]: the distance between two consecutive events.
A session starts after a gap and is extended as long as there are events no farther than the specified distance from each other. Think web browser sessions.

File: docs.feldera.com/docs/sql/table.md (hunk 4 above)
Comment (mihaibudiu) [reply docs.feldera.com/docs/sql/table.md:130]: I will copy an example from the tests

File: sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/streaming/StreamingTests.java (hunk 2 above)
Comment (mihaibudiu) [reply sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/streaming/StreamingTests.java:888]: WATERLINES are the main mechanism used for GC when a collection has LATENESS: a waterline is computed at runtime for each collection where old elements can never be updated. The point here is that a session window start `window_start` can always change, no matter how old, by deleting the last row in the session, which can be recent. So you may see updates to very old values in the column `window_start`. This statement means that the compiler should not build a circuit to compute the waterline for this column. 

Our waterline computing algorithm is conservative, we sometimes fail to compute a waterline where it exists. That's the case for `window_end`. There is a separate issue for that: #2805

Comment (mihaibudiu) [reply sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/streaming/StreamingTests.java:888]: There are quite a few other cases where we do not compute a waterline, which in practice means that we do not GC some queries we could GC: #1850

**Review by mythical-fred — APPROVED** [review]
Rebase refresh of the prior APPROVE. Force-pushed onto today's main; the single commit's diff against its parent is identical to the previously reviewed diff except for a 15-line docs addition to `docs.feldera.com/docs/sql/table.md` describing the `SESSION` window function (syntax, example, `window_start`/`window_end` semantics, `NULL` timestamp/key behaviour). That belongs in the same PR and reads well. Re-approving.
